### PR TITLE
New advanced task provider (+ windows_task refacto & improvement)

### PR DESCRIPTION
**Pull request by Annih**
In order to take advantage of most of the [Windows Task Scheduler features](https://msdn.microsoft.com//windows/desktop/aa383608), this PR adds a new LWRP `windows_advanced_task` that currently handles:
* why_run
* multiple Exec actions, other type of action might be implemented but seems useless or obsolete (mail, showmessage, comhandler)
* multiple triggers
* all trigger types (time, daily,weekly, monthly, monthlydow, event, logon, boot, sessionstatechange, idle, registration) with all their features

The `windows_task` LWRP has been refactored to leverage this implementation.

Feedbacks are greatly appreciated.

README will be updated ASAP

cc. @aboten

### CODE REVIEW

File: providers/advanced_task.rb [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+  true
+end
+
+action :create do
+  if !task_exists? || task_need_update? || new_resource.force
+    validate_logon_info
+    converge_by "creating task #{new_resource}" do
+      folder.RegisterTaskDefinition(new_resource.task_name,
+                                    new_task_object,
+                                    0x6,
```
Comment (jmauro) [inline providers/advanced_task.rb]: Maybe you should explain this value?

File: providers/advanced_task.rb [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+action :update do
+  if task_exists?
+    if task_need_update?
+      validate_logon_info
+
+      converge_by "updating task #{new_resource}" do
+        folder.RegisterTaskDefinition(new_resource.task_name,
+                                      new_task_object,
+                                      0x4,
```
Comment (jmauro) [inline providers/advanced_task.rb]: The same as above.

**Comment by kamaradclimber** [comment]
@tas50 @someara is there anything I can to help in getting this PR merged?

**Comment by Tech356** [comment]
Is this PR being reviewed to be merged?

**Comment by smurawski** [comment]
@Tech356 @Annih @kamaradclimber Looking at this PR later today.  Sorry for the delay.

**Comment by Tech356** [comment]
@smurawski Any updates?

**Comment by Annih** [comment]
Hello @smurawski @mwrock  and other chef windows cookbook maintainers possibly interested in this PR,

1. Is there any way I can update this PR to get it merge? If you have no interest in this provider or plan to implement it differently - using XML, please tell me I'll move the provider in a different cookbook.

2. Would it be better  if `windows_task` does not rely on the new `windows_advanced_task` provider?
Currently the `windows_task` provider has some bugs, lacks many features, and is "hanged quite regularly. It's a real pain to maintain my `windows_task` implementation. So if you don't want this feature, please tell me, I'll just keep the code for the implementation.

Thanks in advance for your answer.

**Comment by mwrock** [comment]
I unfortunately don't have time to review and test in detail soon. I have browsed over this PR a few times since it was submitted and here are some thoughts:
* It does look like a superior implementation to the current task resource.
* I'd be hesitant to have 2 separate implementations/resources and would prefer just one
* ideally, this would actually just go to the core chef client and we would then deprecate the current task

The third point is not trivial and needs thorough testing, which may be why I have avoided this PR :)

**Comment by albeekj** [comment]
Has there been any decision Whether or not this will be merged in to the main cookbook or not?  This would be great to have in chef core, but in the mean time can it be merged so we could utilize it soon?

**Comment by vinyar** [comment]
bump

**Comment by Annih** [comment]
Hello,

I keep this PR open for people interested in this feature, but I personally move this code in a wrapper cookbook to follow windows cookbook releases without having to rebase and maintain the compatibility code.

If someone want to close this issue, let's close it :)

**Comment by mwrock** [comment]
I'm closing but I do like the general approach here and would like to borrow from it when we move this to chef core.

**Comment by Tech356** [comment]
@mwrock Would it be possible to remove the overlap in this code and the current `windows_task` provider and then merge the code? That way the code could be used by the people who want to use it and It wouldn't cause any breaking changes with the existing `windows_task` provider. When the windows task code is moved into chef core, both `windows_task` and `windows_advanced_task` can be deprecated.

**Comment by mwrock** [comment]
Absolutely. If someone wants to take that on, I'd be happy to review the PR.

**Pull request closed without merging**